src/ray/gcs/gcs_server/gcs_health_check_manager.h

```diff
@@ -42,7 +42,7 @@ namespace ray::gcs {
 /// All IO operations happens on the same thread, which is managed by the pass-ed in
 /// [io_service].
 /// TODO (iycheng): Move the GcsHealthCheckManager to ray/common.
-class GcsHealthCheckManager {
+class GcsHealthCheckManager : public std::enable_shared_from_this<GcsHealthCheckManager> {
```

Comment:
We need to make sure `GcsHealthCheckManager` constructor is private so only shared_ptr of it can be created (similar to the retryable_grpc_client):

```
It is permitted to call shared_from_this only on a previously shared object, i.e. on an object managed by [std::shared_ptr](http://en.cppreference.com/w/cpp/memory/shared_ptr)<T>. Otherwise, [std::bad_weak_ptr](https://en.cppreference.com/w/cpp/memory/bad_weak_ptr) is thrown (by the [std::shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr) constructor from a default-constructed weak_this ﻿).
```

Reply:
https://en.cppreference.com/w/cpp/memory/enable_shared_from_this

Reply:
That's a good point, turned into factory function and private constructor.